libflux/src/flux/lib.rs

```diff
@@ -89,13 +90,11 @@ pub unsafe extern "C" fn flux_parse_fb(src_ptr: *const c_char) -> *mut flux_buff
     };
     let r = ast::flatbuffers::serialize(&pkg);
     match r {
-        Ok((vec, offset)) => {
-            let data = &vec[offset..];
-            Box::into_raw(Box::new(flux_buffer_t {
-                data: data.as_ptr(),
-                len: data.len(),
-            }))
-        }
+        Ok((vec, offset)) => Box::into_raw(Box::new(flux_buffer_t {
+            data: vec.as_ptr(),
+            offset,
+            len: vec.len(),
+        })),
         Err(_) => 1 as *mut flux_buffer_t,
```

Comment:
I'm having trouble understanding why the original code worked okay.  Here's what I think the original code was doing:
- taking a slice of the `Vec` that trims the unused first portion
- Then we create a Box that allocates a `flux_buffer_t` on the heap, which contains a pointer to the interior of the `vec`
-  Then the function returns the `Box`.  I guess `Box` will silently convert to a raw pointer, and the caller takes ownership?
- It seems like memory for `vec` will be freed at the end of this function though?

Reply:
The new code also seems to have the problem that you are returning a pointer to a `flux_buffer_t` whose `data` field references memory that is owned by the Vec, that I think will be freed when it goes out of scope.

Does the `unsafe` qualifier at the beginning of the function mean you don't need `unsafe {...}` within the function body?

I think you are right that there is no copy here though, I was wrong about that.

Maybe the right way to do this is:
```rust
let data = Box::into_raw(vec.into_boxed_slice())
Box::into_raw(Box::new(flux_buffer_t {
            data,
            offset,
            len: vec.len(),
        }))
```
I'm not really sure about it though.  There is also `mem::forget` that could be helpful too.